metricflow/naming/object_builder_scheme.py

```diff
@@ -0,0 +1,200 @@
+from __future__ import annotations
+
+import logging
+import re
+from typing import Optional, Sequence
+
+from dbt_semantic_interfaces.call_parameter_sets import ParseWhereFilterException
+from dbt_semantic_interfaces.implementations.filters.where_filter import PydanticWhereFilter
+from dbt_semantic_interfaces.naming.keywords import DUNDER
+from dbt_semantic_interfaces.parsing.where_filter.where_filter_parser import WhereFilterParser
+from dbt_semantic_interfaces.references import EntityReference
+from dbt_semantic_interfaces.type_enums import TimeGranularity
+from dbt_semantic_interfaces.type_enums.date_part import DatePart
+from typing_extensions import override
+
+from metricflow.naming.naming_scheme import QueryItemNamingScheme
+from metricflow.specs.patterns.entity_link_pattern import (
+    DimensionPattern,
+    EntityLinkPattern,
+    EntityLinkPatternParameterSet,
+    ParameterSetField,
+    TimeDimensionPattern,
+)
+from metricflow.specs.patterns.spec_pattern import SpecPattern
+from metricflow.specs.specs import LinkableInstanceSpec, LinkableSpecSet, LinkableSpecSetTransform
+
+logger = logging.getLogger(__name__)
+
+
+class ObjectBuilderNamingScheme(QueryItemNamingScheme):
+    """A naming scheme using a builder syntax like Dimension('metric_time').grain('day')."""
+
+    _NAME_REGEX = re.compile(r"\A(Dimension|TimeDimension|Entity)\(.*\)\Z")
+
+    @override
+    def input_str(self, instance_spec: LinkableInstanceSpec) -> Optional[str]:
+        names = _ObjectBuilderNameTransform().transform(LinkableSpecSet.from_specs((instance_spec,)))
+
+        if len(names) != 1:
+            raise RuntimeError(f"Did not get exactly 1 name from {instance_spec}. Got {names}")
+
+        return names[0]
+
+    @override
+    def spec_pattern(self, input_str: str) -> SpecPattern:
+        if not self.input_str_follows_scheme(input_str):
+            raise ValueError(
```

Comment:
Nice, ValueError. Same question about a custom exception applies, and we should be consistent here.